clang/test/Sema/attr-nonblocking-syntax.cpp

```diff
@@ -0,0 +1,117 @@
+// RUN: %clang_cc1 %s -ast-dump -fblocks | FileCheck %s
+
+// Make sure that the attribute gets parsed and attached to the correct AST elements.
+
+#pragma clang diagnostic ignored "-Wunused-variable"
+
+// =========================================================================================
+// Square brackets, true
+
+namespace square_brackets {
+
+// On the type of the FunctionDecl
+void nl_function() [[clang::nonblocking]];
+// CHECK: FunctionDecl {{.*}} nl_function 'void () __attribute__((clang_nonblocking))'
+
+// On the type of the VarDecl holding a function pointer
+void (*nl_func_a)() [[clang::nonblocking]];
+// CHECK: VarDecl {{.*}} nl_func_a 'void (*)() __attribute__((clang_nonblocking))'
+
+// On the type of the ParmVarDecl of a function parameter
+static void nlReceiver(void (*nl_func)() [[clang::nonblocking]]);
+// CHECK: ParmVarDecl {{.*}} nl_func 'void (*)() __attribute__((clang_nonblocking))'
+
+// As an AttributedType within the nested types of a typedef
+typedef void (*nl_fp_type)() [[clang::nonblocking]];
+// CHECK: TypedefDecl {{.*}} nl_fp_type 'void (*)() __attribute__((clang_nonblocking))'
+using nl_fp_talias = void (*)() [[clang::nonblocking]];
+// CHECK: TypeAliasDecl {{.*}} nl_fp_talias 'void (*)() __attribute__((clang_nonblocking))'
+
+// From a typedef or typealias, on a VarDecl
+nl_fp_type nl_fp_var1;
+// CHECK: VarDecl {{.*}} nl_fp_var1 'nl_fp_type':'void (*)() __attribute__((clang_nonblocking))'
+nl_fp_talias nl_fp_var2;
+// CHECK: VarDecl {{.*}} nl_fp_var2 'nl_fp_talias':'void (*)() __attribute__((clang_nonblocking))'
+
+// On type of a FieldDecl
+struct Struct {
+	void (*nl_func_field)() [[clang::nonblocking]];
+// CHECK: FieldDecl {{.*}} nl_func_field 'void (*)() __attribute__((clang_nonblocking))'
+};
+
+// nonallocating should be subsumed into nonblocking
+void nl1() [[clang::nonblocking]] [[clang::nonallocating]];
+// CHECK: FunctionDecl {{.*}} nl1 'void () __attribute__((clang_nonblocking))'
```

Comment:
This seems like a bad idea for source fidelity reasons; one *effect* subsuming the other in the function type sounds reasonable, but I think both *attributes* should still be present at the AST level.

Reply:
Hm. OK. I can see two ways to accomplish this:

- store both `nonblocking` and `nonallocating` `FunctionEffects` in the `FunctionEffectSet` attached to the `FunctionProtoType`, despite the former constantly trumping the other.
- store only the `nonblocking` effect in the `FunctionEffectSet` but also preserve attributes with `AttributedType` sugar.

Reply:
> * store only the `nonblocking` effect in the `FunctionEffectSet` but also preserve attributes with `AttributedType` sugar.

This is more or less what I had in mind here, yeah; the attributes are spelt in source so they should be preserved; the effects are just computed, so you can do what you want there basically.

Reply:
Right; I suppose that ties back to the fact that our handling of `AttributedType`s isn’t exactly ideal in my experience at least (see [this comment](https://github.com/llvm/llvm-project/pull/84983#issuecomment-2048239663)). As I see it, the second option would be ‘better’ in that that’s literally the reason why `AttributedType`s exist, but if the first option is easier to do then I’d go w/ that for the time being and we can come back to this later as part of a general `AttributedType`s refactor.

Reply:
Now both effects remain present in the `FunctionProtoType`